app/controllers/supplementary/AdditionalInformationController.scala

```diff
@@ -34,27 +39,158 @@ class AdditionalInformationController @Inject()(
   appConfig: AppConfig,
   override val messagesApi: MessagesApi,
   authenticate: AuthAction,
+  errorHandler: ErrorHandler,
   customsCacheService: CustomsCacheService
 )(implicit ec: ExecutionContext)
     extends FrontendController with I18nSupport {
 
   def displayForm(): Action[AnyContent] = authenticate.async { implicit request =>
-    customsCacheService.fetchAndGetEntry[AdditionalInformation](supplementaryCacheId, formId).map {
-      case Some(data) => Ok(additional_information(appConfig, form.fill(data)))
-      case _          => Ok(additional_information(appConfig, form))
+    customsCacheService.fetchAndGetEntry[AdditionalInformationData](supplementaryCacheId, formId).map {
+      case Some(data) => Ok(additional_information(appConfig, form, data.items))
+      case _          => Ok(additional_information(appConfig, form, Seq()))
     }
   }
 
   def saveAdditionalInfo(): Action[AnyContent] = authenticate.async { implicit request =>
-    form
-      .bindFromRequest()
-      .fold(
-        (formWithErrors: Form[AdditionalInformation]) =>
-          Future.successful(Ok(additional_information(appConfig, formWithErrors))),
-        validForm =>
-          customsCacheService.cache[AdditionalInformation](supplementaryCacheId, formId, validForm).map { _ =>
-            Redirect(controllers.supplementary.routes.DocumentsProducedController.displayForm())
+    val boundForm = form.bindFromRequest()
+
+    val actionType =
+      request.body.asFormUrlEncoded.flatMap(_.get("action")).flatMap(_.headOption).getOrElse("Wrong action")
+
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[AdditionalInformationData](supplementaryCacheId, formId)
+      .map(_.getOrElse(AdditionalInformationData(Seq())))
+
+    cachedData.flatMap { cache =>
+      boundForm
+        .fold(
+          (formWithErrors: Form[AdditionalInformation]) =>
+            Future.successful(BadRequest(additional_information(appConfig, formWithErrors, cache.items))),
+          validForm =>
+            actionType match {
+              case "Add"                             => addItem(validForm, cache)
```

Comment:
@MaciejRewera changed the approach of using those cases. Please rebase or update your branch and run tests after it, probably it will fail

Look here and below for action type match: https://github.com/hmrc/customs-declare-exports-frontend/blob/master/app/controllers/supplementary/DeclarationHolderController.scala#L60

Reply:
Thank you @PatrykRudnicki and @MaciejRewera! This has been addressed, branch rebased  and commit amended.